radixdlt-engine/src/main/java/com/radixdlt/constraintmachine/REInstruction.java

```diff
@@ -31,95 +31,119 @@
-		}, REOp.DOWNALL),
-		MSG((byte) 6, (txn, i, b) -> {
-			var length = Byte.toUnsignedInt(b.get());
-			var bytes = new byte[length];
-			b.get(bytes);
-			return bytes;
-		}, REOp.MSG),
-		SIG((byte) 7, (txn, i, b) -> {
-			return RESerializer.deserializeSignature(b);
-		}, REOp.SIG),
-		SYSCALL((byte) 9, (txn, i, b) -> {
-			int bufSize = Byte.toUnsignedInt(b.get());
-			// TODO: Remove buffer copy
-			var callData = new byte[bufSize];
-			b.get(callData);
-			return new CallData(callData);
-		}, REOp.SYSCALL),
-		HEADER((byte) 10, (txn, i, b) -> {
-			int version = b.get();
-			if (version != 0) {
-				throw new DeserializeException("Version must be 0");
+		},
+		VDOWN((byte) 2, REOp.DOWN) {
+			@Override
+			public Object read(Txn txn, int index, ByteBuffer buf, SubstateDeserialization d) throws DeserializeException {
+				int pos = buf.position();
+				var p = d.deserialize(buf);
+				int length = buf.position() - pos;
+				var b = ByteBuffer.wrap(buf.array(), pos, length);
+				return Substate.create(p, SubstateId.ofVirtualSubstate(b));
 			}
-			var flags = b.get();
-			if ((flags & 0xe) != 0) {
-				throw new DeserializeException("Invalid flags");
+		},
+		VDOWNARG((byte) 3, REOp.DOWN) {
+			@Override
+			public Object read(Txn txn, int index, ByteBuffer buf, SubstateDeserialization d) throws DeserializeException {
+				int pos = buf.position();
+				var p = d.deserialize(buf);
+				int length = buf.position() - pos;
+				var b = ByteBuffer.wrap(buf.array(), pos, length);
+				var argLength = buf.get();
+				var arg = new byte[Byte.toUnsignedInt(argLength)];
+				buf.get(arg);
+				return Pair.of(Substate.create(p, SubstateId.ofVirtualSubstate(b)), arg);
 			}
-			return (flags & 0x1) == 1;
-		}, REOp.HEADER),
-		END((byte) 0, (txn, i, b) -> null, REOp.END);
-
+		},
+		DOWN((byte) 4, REOp.DOWN) {
+			@Override
+			public Object read(Txn txn, int index, ByteBuffer buf, SubstateDeserialization d) throws DeserializeException {
+				return SubstateId.fromBuffer(buf);
+			}
+		},
+		LDOWN((byte) 5, REOp.DOWN) {
+			@Override
+			public Object read(Txn txn, int i, ByteBuffer b, SubstateDeserialization d) throws DeserializeException {
+				var index = b.getInt();
+				if (index < 0 || index >= i) {
+					throw new DeserializeException("Bad local index: " + index);
+				}
+				return SubstateId.ofSubstate(txn.getId(), index);
+			}
+		},
+		DOWNALL((byte) 8, REOp.DOWNALL) {
+			@Override
+			public Object read(Txn txn, int i, ByteBuffer b, SubstateDeserialization d) throws DeserializeException {
+				var classId = b.get();
+				return d.byteToClass(classId); // Just to check to make sure classId exists
+			}
+		},
+		MSG((byte) 6, REOp.MSG) {
+			@Override
+			public Object read(Txn txn, int i, ByteBuffer b, SubstateDeserialization d) throws DeserializeException {
+				var length = Byte.toUnsignedInt(b.get());
+				var bytes = new byte[length];
+				b.get(bytes);
```

Comment:
We may want to create a reusable method for decoding variable length byte array. It seems to be commonly used.